### PR TITLE
Chainlink develop

### LINKED ISSUE #5: No interaction with ChainLink contract
**Issue**
In the contract we have a ```report()``` function that checks the activity data for a commitment via Chainlink and could update the ```reportedValue``` field of a Commitment. Currently, it simply returns ```true``` and the value is not updated

**Example**
```
    function report() public returns (bool) {
        // get activity data from oracle
        // record activity data in commitments
        return true;
    }
```

**Expected outcome**
When the report function is called:
- Interaction with ChainLink contract to get user data
- Update reportedValue of Commitment

### PR DESCRIPTION
**Pull request by bitbeckers**
- Management of activity state methods
- Keeping track of slashed balances
- Updated committerBalance methods and added slash method
- Refactoring of field/mapping names (e.g. committerBalance to totalCommitterBalance & activityList to activityKeyList)
- Updated tests
- Additional TODOs

### CODE REVIEW

File: contracts/SinglePlayerCommit.sol
Comment (bitbeckers) [inline contracts/SinglePlayerCommit.sol:254]: First check on commitment.exists. Call from storage returns default object, so require() checks could lead to some null/0 comparisons

File: contracts/SinglePlayerCommit.sol
Comment (bitbeckers) [inline contracts/SinglePlayerCommit.sol]: Faulty logic. _changeCommitterBalance automatically slashes from msg.sender's balance. But should be based on committer address

File: contracts/SinglePlayerCommit.sol
Comment (spengrah) [inline contracts/SinglePlayerCommit.sol:62]: efficiency police: do we need a Deposit event? Could we get the same information from the event emitted by the token transfer?

File: contracts/SinglePlayerCommit.sol
Comment (spengrah) [inline contracts/SinglePlayerCommit.sol]: I think this needs to be `<=` not greater than or equal. The available amount must cover the withdrawal amount.

File: contracts/SinglePlayerCommit.sol
Comment (spengrah) [inline contracts/SinglePlayerCommit.sol:63]: efficiency police: do we need a Withdrawal event? Could we get the same information from the event emitted by the token transfer?

File: contracts/SinglePlayerCommit.sol
Comment (spengrah) [inline contracts/SinglePlayerCommit.sol:33]: what does this give us that allowed doesn't?

File: contracts/SinglePlayerCommit.sol
Comment (spengrah) [inline contracts/SinglePlayerCommit.sol:426]: whats the difference between disabling and disallowing an activity as in `updateActivityAllowed`?

put another way, what is the benefit of deleting an activity vs. disallowing it?

File: contracts/SinglePlayerCommit.sol
Comment (spengrah) [inline contracts/SinglePlayerCommit.sol]: I think this should be `>=`

File: contracts/SinglePlayerCommit.sol
Comment (spengrah) [inline contracts/SinglePlayerCommit.sol]: let's factor the common logic from `processCommitment` and `processCommitmentUser` out into its own (internal?) function

File: contracts/SinglePlayerCommit.sol
Comment (spengrah) [inline contracts/SinglePlayerCommit.sol]: use safemath here

File: contracts/SinglePlayerCommit.sol
Comment (spengrah) [inline contracts/SinglePlayerCommit.sol:296]: do the transfer after updating `slashedBalance` (re-entry protection)

File: contracts/SinglePlayerCommit.sol
Comment (spengrah) [inline contracts/SinglePlayerCommit.sol]: use safemath here

File: contracts/SinglePlayerCommit.sol
Comment (spengrah) [reply contracts/SinglePlayerCommit.sol]: maybe simplest to revise `_changeCommitterBalance` to take an `address committer` argument?

File: contracts/SinglePlayerCommit.sol
Comment (spengrah) [inline contracts/SinglePlayerCommit.sol]: activityKeyList

File: contracts/SinglePlayerCommit.sol
Comment (spengrah) [inline contracts/SinglePlayerCommit.sol]: activityKeyList

**Review by spengrah — CHANGES_REQUESTED** [review]

File: contracts/SinglePlayerCommit.sol
Comment (bitbeckers) [reply contracts/SinglePlayerCommit.sol:33]: It can exist and not be allowed, leaving room for already active commitments to be resolved while migrating away from a certain activity by not letting it be available for newer commitments

File: contracts/SinglePlayerCommit.sol
Comment (bitbeckers) [reply contracts/SinglePlayerCommit.sol:426]: See other discussion

File: contracts/SinglePlayerCommit.sol
Comment (bitbeckers) [reply contracts/SinglePlayerCommit.sol]: agreed

File: contracts/SinglePlayerCommit.sol
Comment (bitbeckers) [reply contracts/SinglePlayerCommit.sol]: This entails that processing commitment has 2 possible subsequent calls, first to the common logic, than to the slashin function. I believe you mentioned something about not adding to much method calls? That's why I went for this approach, so are we sure?

File: contracts/SinglePlayerCommit.sol
Comment (bitbeckers) [reply contracts/SinglePlayerCommit.sol]: Done

File: contracts/SinglePlayerCommit.sol
Comment (bitbeckers) [reply contracts/SinglePlayerCommit.sol:296]: Done

File: contracts/SinglePlayerCommit.sol
Comment (bitbeckers) [reply contracts/SinglePlayerCommit.sol]: Done

File: contracts/SinglePlayerCommit.sol
Comment (bitbeckers) [reply contracts/SinglePlayerCommit.sol]: Sharp, done

File: contracts/SinglePlayerCommit.sol
Comment (bitbeckers) [reply contracts/SinglePlayerCommit.sol]: Done

File: contracts/SinglePlayerCommit.sol
Comment (bitbeckers) [reply contracts/SinglePlayerCommit.sol:63]: There are more things to review on the Events, like adding error events. To keep withing scope, do we want to revise them now?

File: contracts/SinglePlayerCommit.sol
Comment (bitbeckers) [reply contracts/SinglePlayerCommit.sol]: Oopsie. Fixed

File: contracts/SinglePlayerCommit.sol
Comment (bitbeckers) [reply contracts/SinglePlayerCommit.sol]: I stand corrected: https://docs.soliditylang.org/en/v0.5.4/control-structures.html#internal-function-calls

So I'll add it as an internal method

File: contracts/SinglePlayerCommit.sol
Comment (spengrah) [reply contracts/SinglePlayerCommit.sol:33]: ah, good point!

File: contracts/SinglePlayerCommit.sol
Comment (spengrah) [reply contracts/SinglePlayerCommit.sol]: cool, thanks

Comment (bitbeckers) [reply contracts/SinglePlayerCommit.sol]: done

File: contracts/SinglePlayerCommit.sol
Comment (bitbeckers) [reply contracts/SinglePlayerCommit.sol:254]: done

File: contracts/SinglePlayerCommit.sol
Comment (bitbeckers) [reply contracts/SinglePlayerCommit.sol]: done

File: README.md
Comment (bitbeckers) [inline README.md]: Update before merge

**Comment by spengrah** [comment]
Task linked: [CU-gwyjqp merge contract into master](https://app.clickup.com/t/gwyjqp)

File: contracts/SinglePlayerCommit.sol
Comment (spengrah) [reply contracts/SinglePlayerCommit.sol]: nice! 

minor naming change: `_settleCommitment` instead of `_commitmentSettlement` is more of a verb/method/function

**Review by spengrah — APPROVED** [review]
looks great!

File: contracts/SinglePlayerCommit.sol
Comment (bitbeckers) [reply contracts/SinglePlayerCommit.sol:63]: Created an Issue #16

File: contracts/SinglePlayerCommit.sol
Comment (bitbeckers) [reply contracts/SinglePlayerCommit.sol:62]: Created issue #16

File: README.md
Comment (bitbeckers) [reply README.md]: done

**Pull request merged**
